lib/Ryu/Async.pm

```diff
@@ -381,17 +380,20 @@ sub udp_client {
     })->on_fail(sub {
         $log->errorf("UDP client failed to connect - %s", join ',', @_);
     });
-    $sink->source->each(sub {
-        my $payload = $_;
+    $sink->source->each($sink->$curry::weak( sub {
+        my ($sink_obj, $payload) = @_;
         $f->on_done(sub {
             try {
                 $log->tracef("Sending [%s] to %s", $payload, $uri);
                 $client->send($payload, undef, "$host:$port");
             } catch {
-                $log->errorf("Exception when sending: %s", $@);
+                my $err = "Exception when sending to $host:$port - %s" . $@;
+                $log->errorf($err);
```

Comment:
I think no need to log if there's an error raised - instead, log an error if either Future ->is_ready, otherwise mark them as failed.